File: swift/src/Filter.swift
// Filter.swift - Clean/smudge filter system
// Parity: libgit2 src/libgit2/filter.c, crlf.c, ident.c

import Foundation

/// Direction of filtering.
/// Parity: git_filter_mode_t
public enum FilterMode {
    /// Working directory → ODB (clean)
    case toOdb
    /// ODB → working directory (smudge)
    case toWorktree
}

/// Metadata about the file being filtered.
/// Parity: git_filter_source
public struct FilterSource {
    public let path: String
    public let mode: FilterMode
    public let oid: OID?

    public init(path: String, mode: FilterMode, oid: OID? = nil) {
        self.path = path
        self.mode = mode
        self.oid = oid
    }
}

/// Result of applying a filter.
public enum FilterResult {
    case applied(Data)
    case passthrough
}

/// A single filter implementation.
public protocol FilterProtocol {
    var name: String { get }
    func check(source: FilterSource, attrs: Attributes) -> Bool
    func apply(input: Data, source: FilterSource) -> FilterResult
}

/// A chain of filters to apply to a file.
/// Parity: git_filter_list
public class FilterList {
    private var filters: [FilterProtocol] = []
    public let source: FilterSource

    init(source: FilterSource) {
        self.source = source
    }

    /// Load the applicable filters for a path in a repository.
    /// Parity: git_filter_list_load
    public static func load(
        gitDir: String,
        workdir: String?,
        path: String,
        mode: FilterMode,
        oid: OID? = nil
    ) -> FilterList {
        let attrs = Attributes.loadForRepo(gitDir: gitDir, workdir: workdir)
        let source = FilterSource(path: path, mode: mode, oid: oid)
        let list = FilterList(source: source)

        let crlf = CrlfFilter(gitDir: gitDir)
        let ident = IdentFilter()

        switch mode {
        case .toWorktree:
            // Smudge: CRLF(0) → Ident(100)
            if crlf.check(source: source, attrs: attrs) {
                list.filters.append(crlf)
            }
            if ident.check(source: source, attrs: attrs) {
                list.filters.append(ident)
            }
        case .toOdb:
            // Clean: Ident(100) → CRLF(0) (reverse order)
            if ident.check(source: source, attrs: attrs) {
                list.filters.append(ident)
            }
            if crlf.check(source: source, attrs: attrs) {
                list.filters.append(crlf)
            }
        }

        return list
    }

    /// Apply all filters in the chain to the input data.
    /// Parity: git_filter_list_apply_to_buffer
    public func apply(_ input: Data) -> Data {
        var data = input
        for filter in filters {
            switch filter.apply(input: data, source: source) {
            case .applied(let output):
                data = output
            case .passthrough:
                break
            }
        }
        return data
    }

    /// Number of active filters.
    public var count: Int { filters.count }

    /// Whether the filter list is empty.
    public var isEmpty: Bool { filters.isEmpty }

    /// Check if a named filter is in the list.
    public func contains(_ name: String) -> Bool {
        filters.contains { $0.name == name }
    }
}

// MARK: - CRLF Filter (priority 0)
// Parity: libgit2 src/libgit2/crlf.c

/// Resolved CRLF action.
private enum CrlfAction {
    case none
    case crlfToLf
    case lfToCrlf
    case auto
}

/// End-of-line style.
private enum EolStyle {
    case lf
    case crlf
    case native
}

/// CRLF / text / eol filter.
public class CrlfFilter: FilterProtocol {
    private let autoCrlf: String?
    private let coreEol: String?

    public init(gitDir: String) {
        let configPath = (gitDir as NSString).appendingPathComponent("config")
        let config = (try? Config.load(from: configPath)) ?? Config()
        self.autoCrlf = config.get(section: "core", key: "autocrlf")
        self.coreEol = config.get(section: "core", key: "eol")
    }

    init(autoCrlf: String?, coreEol: String?) {
        self.autoCrlf = autoCrlf
        self.coreEol = coreEol
    }

    public var name: String { "crlf" }

    public func check(source: FilterSource, attrs: Attributes) -> Bool {
        let action = resolveAction(attrs: attrs, path: source.path, mode: source.mode)
        return action != .none
    }

    public func apply(input: Data, source: FilterSource) -> FilterResult {
        if isBinary(input) {
            return .passthrough
        }
        switch source.mode {
        case .toOdb:
            return crlfToLf(input)
        case .toWorktree:
            return lfToCrlf(input)
        }
    }

    private func resolveAction(attrs: Attributes, path: String, mode: FilterMode) -> CrlfAction {
        let textAttr = attrs.get(path, attr: "text")
        let crlfAttr = attrs.get(path, attr: "crlf")
        let eolAttr = attrs.get(path, attr: "eol")

        // text attribute takes priority
        var isText: Bool? = nil
        switch textAttr {
        case .set:
            isText = true
        case .unset:
            isText = false
        case .value(let v) where v == "auto":
            return .auto
        default:
            break
        }

        // Fall back to crlf attribute
        if isText == nil {
            switch crlfAttr {
            case .set:
                isText = true
            case .unset:
                isText = false
            default:
                break
            }
        }

        // Fall back to core.autocrlf config
        if isText == nil {
            switch autoCrlf {
            case "true", "input":
                isText = true
            default:
                break
            }
        }

        guard isText == true else {
            return .none
        }

        let outputEol = resolveEol(eolAttr: eolAttr, mode: mode)

        switch mode {
        case .toOdb:
            return .crlfToLf
        case .toWorktree:
            switch outputEol {
            case .crlf:
                return .lfToCrlf
            case .lf:
                return .none
            case .native:
                #if os(Windows)
                return .lfToCrlf
                #else
                return .none
                #endif
            }
        }
    }

    private func resolveEol(eolAttr: AttrValue?, mode: FilterMode) -> EolStyle {
        if case .value(let v) = eolAttr {
            switch v {
            case "lf": return .lf
            case "crlf": return .crlf
            default: return .native
            }
        }

        if mode == .toOdb, autoCrlf == "input" {
            return .lf
        }

        switch coreEol {
        case "lf": return .lf
        case "crlf": return .crlf
        default: return .native
        }
    }
}

// MARK: - Ident Filter (priority 100)
// Parity: libgit2 src/libgit2/ident.c

/// $Id$ expansion/contraction filter.
public class IdentFilter: FilterProtocol {
    public init() {}

    public var name: String { "ident" }

    public func check(source: FilterSource, attrs: Attributes) -> Bool {
        attrs.get(source.path, attr: "ident") == .set
    }

    public func apply(input: Data, source: FilterSource) -> FilterResult {
        if isBinary(input) {
            return .passthrough
        }
        switch source.mode {
        case .toWorktree:
            return identSmudge(input, oid: source.oid)
        case .toOdb:
            return identClean(input)
        }
    }
}

// MARK: - Internal Helpers

/// Convert CRLF to LF (clean direction).
func crlfToLf(_ input: Data) -> FilterResult {
    let bytes = Array(input)
    guard bytes.contains(where: { $0 == 0x0D }) else {
        // Quick check: no CR at all
        return .passthrough
    }

    var hasCrlf = false
    for i in 0..<(bytes.count - 1) {
        if bytes[i] == 0x0D && bytes[i + 1] == 0x0A {
            hasCrlf = true
            break
        }
    }
    guard hasCrlf else {
        return .passthrough
    }

    var output = Data()
    output.reserveCapacity(bytes.count)
    var i = 0
    while i < bytes.count {
        if i + 1 < bytes.count && bytes[i] == 0x0D && bytes[i + 1] == 0x0A {
            output.append(0x0A)
            i += 2
        } else {
            output.append(bytes[i])
            i += 1
        }
    }
    return .applied(output)
}

/// Convert LF to CRLF (smudge direction).
func lfToCrlf(_ input: Data) -> FilterResult {
    let bytes = Array(input)
    var hasBareLf = false
    for i in 0..<bytes.count {
        if bytes[i] == 0x0A && (i == 0 || bytes[i - 1] != 0x0D) {
            hasBareLf = true
            break
        }
    }
    guard hasBareLf else {
        return .passthrough
    }

    var output = Data()
    output.reserveCapacity(bytes.count + bytes.count / 10)
    for i in 0..<bytes.count {
        if bytes[i] == 0x0A && (i == 0 || bytes[i - 1] != 0x0D) {
            output.append(0x0D)
        }
        output.append(bytes[i])
    }
    return .applied(output)
}

/// Simple binary detection: check for NUL bytes in the first 8000 bytes.
func isBinary(_ data: Data) -> Bool {
    let checkLen = min(data.count, 8000)
    return data.prefix(checkLen).contains(0)
}

/// Smudge: Replace `$Id$` with `$Id: <hex> $`
func identSmudge(_ input: Data, oid: OID?) -> FilterResult {
    guard let oid = oid else { return .passthrough }
    guard let str = String(data: input, encoding: .utf8) else { return .passthrough }

    let needle = "$Id$"
    let replacement = "$Id: \(oid.hex) $"

    guard str.contains(needle) else { return .passthrough }

    let result = str.replacingOccurrences(of: needle, with: replacement)
    return .applied(Data(result.utf8))
}

/// Clean: Replace `$Id: <anything> $` back to `$Id$`
func identClean(_ input: Data) -> FilterResult {
    guard let str = String(data: input, encoding: .utf8) else { return .passthrough }
    guard str.contains("$Id:") else { return .passthrough }

    var output = ""
    var remaining = str[str.startIndex...]

    while let startRange = remaining.range(of: "$Id:") {
        output += remaining[remaining.startIndex..<startRange.lowerBound]
        let afterStart = remaining[startRange.upperBound...]

        if let endRange = afterStart.range(of: "$") {
            let content = afterStart[afterStart.startIndex..<endRange.lowerBound]
            if !content.contains("\n") {
                output += "$Id$"
                remaining = afterStart[endRange.upperBound...]
            } else {
                output += "$Id:"
                remaining = afterStart
            }
        } else {
            output += remaining[startRange.lowerBound...]
            remaining = remaining[remaining.endIndex...]
            break
        }
    }

    output += remaining

    let resultData = Data(output.utf8)
    if resultData == input {
        return .passthrough
    }
    return .applied(resultData)
}

File: rust/src/filter.rs
//! Clean/smudge filter system
//! Parity: libgit2 src/libgit2/filter.c, crlf.c, ident.c

use std::path::Path;

use crate::attributes::{Attributes, AttrValue};
use crate::config::Config;
use crate::oid::OID;

/// Direction of filtering.
/// Parity: git_filter_mode_t
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FilterMode {
    /// Working directory → ODB (clean)
    ToOdb,
    /// ODB → working directory (smudge)
    ToWorktree,
}

/// Metadata about the file being filtered.
/// Parity: git_filter_source
#[derive(Debug, Clone)]
pub struct FilterSource {
    pub path: String,
    pub mode: FilterMode,
    pub oid: Option<OID>,
}

/// Result of applying a filter.
#[derive(Debug)]
pub enum FilterResult {
    Applied(Vec<u8>),
    Passthrough,
}

/// A single filter implementation.
pub trait Filter: std::fmt::Debug {
    fn name(&self) -> &str;
    /// Check if this filter applies to the given source.
    /// Returns true if the filter should be applied.
    fn check(&self, source: &FilterSource, attrs: &Attributes) -> bool;
    /// Apply the filter to the input data.
    fn apply(&self, input: &[u8], source: &FilterSource) -> FilterResult;
}

/// A chain of filters to apply to a file.
/// Parity: git_filter_list
#[derive(Debug)]
pub struct FilterList {
    filters: Vec<Box<dyn Filter>>,
    pub source: FilterSource,
}

impl FilterList {
    /// Load the applicable filters for a path in a repository.
    /// Parity: git_filter_list_load
    pub fn load(
        git_dir: &Path,
        workdir: Option<&Path>,
        path: &str,
        mode: FilterMode,
        oid: Option<OID>,
    ) -> Self {
        let attrs = Attributes::load_for_repo(git_dir, workdir);
        let source = FilterSource {
            path: path.to_string(),
            mode,
            oid,
        };
        let mut filters: Vec<Box<dyn Filter>> = Vec::new();

        // Built-in filters in priority order (lower priority first for smudge)
        let crlf = CrlfFilter::new(git_dir);
        let ident = IdentFilter;

        match mode {
            FilterMode::ToWorktree => {
                // Smudge: CRLF(0) → Ident(100)
                if crlf.check(&source, &attrs) {
                    filters.push(Box::new(crlf));
                }
                if ident.check(&source, &attrs) {
                    filters.push(Box::new(ident));
                }
            }
            FilterMode::ToOdb => {
                // Clean: Ident(100) → CRLF(0) (reverse order)
                if ident.check(&source, &attrs) {
                    filters.push(Box::new(ident));
                }
                if crlf.check(&source, &attrs) {
                    filters.push(Box::new(crlf));
                }
            }
        }

        FilterList { filters, source }
    }

    /// Apply all filters in the chain to the input data.
    /// Parity: git_filter_list_apply_to_buffer
    pub fn apply(&self, input: &[u8]) -> Vec<u8> {
        let mut data = input.to_vec();
        for filter in &self.filters {
            match filter.apply(&data, &self.source) {
                FilterResult::Applied(output) => data = output,
                FilterResult::Passthrough => {}
            }
        }
        data
    }

    /// Number of active filters.
    pub fn len(&self) -> usize {
        self.filters.len()
    }

    /// Whether the filter list is empty.
    pub fn is_empty(&self) -> bool {
        self.filters.is_empty()
    }

    /// Check if a named filter is in the list.
    pub fn contains(&self, name: &str) -> bool {
        self.filters.iter().any(|f| f.name() == name)
    }
}

// ── CRLF Filter (priority 0) ──
// Parity: libgit2 src/libgit2/crlf.c

/// Resolved CRLF action.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum CrlfAction {
    None,
    CrlfToLf,
    LfToCrlf,
    Auto,
}

/// End-of-line style.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum EolStyle {
    Lf,
    Crlf,
    Native,
}

/// CRLF / text / eol filter.
#[derive(Debug)]
pub struct CrlfFilter {
    auto_crlf: Option<String>,
    core_eol: Option<String>,
}

impl CrlfFilter {
    pub fn new(git_dir: &Path) -> Self {
        let config_path = git_dir.join("config");
        let config = Config::load(&config_path).unwrap_or_else(|_| Config::new());
        let auto_crlf = config.get("core", "autocrlf").map(|s| s.to_string());
        let core_eol = config.get("core", "eol").map(|s| s.to_string());
        CrlfFilter { auto_crlf, core_eol }
    }

    fn resolve_action(&self, attrs: &Attributes, path: &str, mode: FilterMode) -> CrlfAction {
        let text_attr = attrs.get(path, "text");
        let crlf_attr = attrs.get(path, "crlf");
        let eol_attr = attrs.get(path, "eol");

        // text attribute takes priority
        let is_text = match text_attr {
            Some(AttrValue::Set) => Some(true),
            Some(AttrValue::Unset) => Some(false),
            Some(AttrValue::Value(v)) if v == "auto" => {
                return CrlfAction::Auto;
            }
            _ => None,
        };

        // Fall back to crlf attribute
        let is_text = is_text.or_else(|| match crlf_attr {
            Some(AttrValue::Set) => Some(true),
            Some(AttrValue::Unset) => Some(false),
            _ => None,
        });

        // Fall back to core.autocrlf config
        let is_text = is_text.or_else(|| match self.auto_crlf.as_deref() {
            Some("true") => Some(true),
            Some("input") => Some(true),
            _ => None,
        });

        let is_text = match is_text {
            Some(true) => true,
            Some(false) => return CrlfAction::None,
            None => return CrlfAction::None,
        };

        if !is_text {
            return CrlfAction::None;
        }

        // Determine output eol
        let output_eol = self.resolve_eol(eol_attr, mode);

        match mode {
            FilterMode::ToOdb => CrlfAction::CrlfToLf,
            FilterMode::ToWorktree => match output_eol {
                EolStyle::Crlf => CrlfAction::LfToCrlf,
                EolStyle::Lf => CrlfAction::None,
                EolStyle::Native => {
                    if cfg!(windows) {
                        CrlfAction::LfToCrlf
                    } else {
                        CrlfAction::None
                    }
                }
            },
        }
    }

    fn resolve_eol(&self, eol_attr: Option<&AttrValue>, mode: FilterMode) -> EolStyle {
        // eol attribute overrides everything
        if let Some(AttrValue::Value(v)) = eol_attr {
            return match v.as_str() {
                "lf" => EolStyle::Lf,
                "crlf" => EolStyle::Crlf,
                _ => EolStyle::Native,
            };
        }

        // core.autocrlf=input means always LF on clean
        if mode == FilterMode::ToOdb {
            if let Some("input") = self.auto_crlf.as_deref() {
                return EolStyle::Lf;
            }
        }

        // core.eol config
        match self.core_eol.as_deref() {
            Some("lf") => EolStyle::Lf,
            Some("crlf") => EolStyle::Crlf,
            _ => EolStyle::Native,
        }
    }
}

impl Filter for CrlfFilter {
    fn name(&self) -> &str {
        "crlf"
    }

    fn check(&self, source: &FilterSource, attrs: &Attributes) -> bool {
        let action = self.resolve_action(attrs, &source.path, source.mode);
        action != CrlfAction::None
    }

    fn apply(&self, input: &[u8], source: &FilterSource) -> FilterResult {
        // Skip binary content
        if is_binary(input) {
            return FilterResult::Passthrough;
        }

        match source.mode {
            FilterMode::ToOdb => {
                // Clean: CRLF → LF
                crlf_to_lf(input)
            }
            FilterMode::ToWorktree => {
                // Smudge: LF → CRLF (only if needed)
                lf_to_crlf(input)
            }
        }
    }
}

/// Convert CRLF to LF (clean direction).
fn crlf_to_lf(input: &[u8]) -> FilterResult {
    if !input.windows(2).any(|w| w == b"\r\n") {
        return FilterResult::Passthrough;
    }

    let mut output = Vec::with_capacity(input.len());
    let mut i = 0;
    while i < input.len() {
        if i + 1 < input.len() && input[i] == b'\r' && input[i + 1] == b'\n' {
            output.push(b'\n');
            i += 2;
        } else {
            output.push(input[i]);
            i += 1;
        }
    }
    FilterResult::Applied(output)
}

/// Convert LF to CRLF (smudge direction).
fn lf_to_crlf(input: &[u8]) -> FilterResult {
    // Check if there's anything to do
    let has_bare_lf = input.windows(1).enumerate().any(|(i, w)| {
        w[0] == b'\n' && (i == 0 || input[i - 1] != b'\r')
    });

    if !has_bare_lf {
        return FilterResult::Passthrough;
    }

    let mut output = Vec::with_capacity(input.len() + input.len() / 10);
    for (i, &byte) in input.iter().enumerate() {
        if byte == b'\n' && (i == 0 || input[i - 1] != b'\r') {
            output.push(b'\r');
        }
        output.push(byte);
    }
    FilterResult::Applied(output)
}

/// Simple binary detection: check for NUL bytes in the first 8000 bytes.
/// Parity: libgit2 src/libgit2/filter.h GIT_FILTER_BYTES_TO_CHECK_NUL
fn is_binary(data: &[u8]) -> bool {
    let check_len = data.len().min(8000);
    data[..check_len].contains(&0)
}

// ── Ident Filter (priority 100) ──
// Parity: libgit2 src/libgit2/ident.c

/// $Id$ expansion/contraction filter.
#[derive(Debug)]
pub struct IdentFilter;

impl Filter for IdentFilter {
    fn name(&self) -> &str {
        "ident"
    }

    fn check(&self, source: &FilterSource, attrs: &Attributes) -> bool {
        matches!(attrs.get(&source.path, "ident"), Some(AttrValue::Set))
    }

    fn apply(&self, input: &[u8], source: &FilterSource) -> FilterResult {
        if is_binary(input) {
            return FilterResult::Passthrough;
        }

        match source.mode {
            FilterMode::ToWorktree => ident_smudge(input, source.oid.as_ref()),
            FilterMode::ToOdb => ident_clean(input),
        }
    }
}

/// Smudge: Replace `$Id$` with `$Id: <hex> $`
fn ident_smudge(input: &[u8], oid: Option<&OID>) -> FilterResult {
    let oid = match oid {
        Some(o) => o,
        None => return FilterResult::Passthrough,
    };

    let needle = b"$Id$";
    let replacement = format!("$Id: {} $", oid.hex());

    if let Some(output) = replace_bytes(input, needle, replacement.as_bytes()) {
        FilterResult::Applied(output)
    } else {
        FilterResult::Passthrough
    }
}

/// Clean: Replace `$Id: <anything> $` back to `$Id$`
fn ident_clean(input: &[u8]) -> FilterResult {
    let start_marker = b"$Id:";
    let end_marker = b"$";

    let input_str = match std::str::from_utf8(input) {
        Ok(s) => s,
        Err(_) => return FilterResult::Passthrough,
    };

    if !input_str.contains("$Id:") {
        return FilterResult::Passthrough;
    }

    let mut output = String::with_capacity(input_str.len());
    let mut remaining = input_str;

    while let Some(start_pos) = remaining.find("$Id:") {
        output.push_str(&remaining[..start_pos]);

        let after_start = &remaining[start_pos + start_marker.len()..];
        if let Some(end_pos) = after_start.find(std::str::from_utf8(end_marker).unwrap()) {
            // Check that the content between $Id: and $ doesn't contain newlines
            let content = &after_start[..end_pos];
            if !content.contains('\n') {
                output.push_str("$Id$");
                remaining = &after_start[end_pos + 1..];
            } else {
                output.push_str("$Id:");
                remaining = after_start;
            }
        } else {
            output.push_str(&remaining[start_pos..]);
            remaining = "";
            break;
        }
    }

    output.push_str(remaining);

    if output.as_bytes() == input {
        FilterResult::Passthrough
    } else {
        FilterResult::Applied(output.into_bytes())
    }
}

/// Replace all occurrences of `needle` in `haystack` with `replacement`.
fn replace_bytes(haystack: &[u8], needle: &[u8], replacement: &[u8]) -> Option<Vec<u8>> {
    if needle.is_empty() || haystack.len() < needle.len() {
        return None;
    }

    let mut found = false;
    let mut output = Vec::with_capacity(haystack.len());
    let mut i = 0;

    while i <= haystack.len() - needle.len() {
        if &haystack[i..i + needle.len()] == needle {
            output.extend_from_slice(replacement);
            i += needle.len();
            found = true;
        } else {
            output.push(haystack[i]);
            i += 1;
        }
    }

    // Append remaining bytes
    output.extend_from_slice(&haystack[i..]);

    if found {
        Some(output)
    } else {
        None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;

    fn test_tmp(name: &str) -> PathBuf {
        let tmp = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../tmp").join(name);
        let _ = fs::remove_dir_all(&tmp);
        fs::create_dir_all(tmp.join(".git/info")).unwrap();
        fs::create_dir_all(tmp.join(".git/objects")).unwrap();
        fs::create_dir_all(tmp.join(".git/refs")).unwrap();
        tmp
    }

    #[test]
    fn test_crlf_to_lf_clean() {
        let result = crlf_to_lf(b"hello\r\nworld\r\n");
        match result {
            FilterResult::Applied(data) => assert_eq!(data, b"hello\nworld\n"),
            FilterResult::Passthrough => panic!("expected Applied"),
        }
    }

    #[test]
    fn test_crlf_to_lf_no_crlf() {
        let result = crlf_to_lf(b"hello\nworld\n");
        assert!(matches!(result, FilterResult::Passthrough));
    }

    #[test]
    fn test_lf_to_crlf_smudge() {
        let result = lf_to_crlf(b"hello\nworld\n");
        match result {
            FilterResult::Applied(data) => assert_eq!(data, b"hello\r\nworld\r\n"),
            FilterResult::Passthrough => panic!("expected Applied"),
        }
    }

    #[test]
    fn test_lf_to_crlf_already_crlf() {
        let result = lf_to_crlf(b"hello\r\nworld\r\n");
        assert!(matches!(result, FilterResult::Passthrough));
    }

    #[test]
    fn test_ident_smudge() {
        let oid = OID::from_hex("aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d").unwrap();
        let source = FilterSource {
            path: "test.txt".to_string(),
            mode: FilterMode::ToWorktree,
            oid: Some(oid.clone()),
        };
        let result = IdentFilter.apply(b"Version: $Id$\n", &source);
        match result {
            FilterResult::Applied(data) => {
                let s = std::str::from_utf8(&data).unwrap();
                assert!(s.contains("$Id: aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d $"));
            }
            FilterResult::Passthrough => panic!("expected Applied"),
        }
    }

    #[test]
    fn test_ident_clean() {
        let source = FilterSource {
            path: "test.txt".to_string(),
            mode: FilterMode::ToOdb,
            oid: None,
        };
        let input = b"Version: $Id: aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d $\n";
        let result = IdentFilter.apply(input, &source);
        match result {
            FilterResult::Applied(data) => {
                assert_eq!(data, b"Version: $Id$\n");
            }
            FilterResult::Passthrough => panic!("expected Applied"),
        }
    }

    #[test]
    fn test_ident_no_marker() {
        let source = FilterSource {
            path: "test.txt".to_string(),
            mode: FilterMode::ToWorktree,
            oid: Some(OID::from_hex("aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d").unwrap()),
        };
        let result = IdentFilter.apply(b"no markers here\n", &source);
        assert!(matches!(result, FilterResult::Passthrough));
    }

    #[test]
    fn test_binary_detection() {
        assert!(is_binary(b"hello\x00world"));
        assert!(!is_binary(b"hello world"));
    }

    #[test]
    fn test_binary_skipped_by_crlf() {
        let source = FilterSource {
            path: "test.bin".to_string(),
            mode: FilterMode::ToOdb,
            oid: None,
        };
        let input = b"hello\r\n\x00world";
        let crlf = CrlfFilter {
            auto_crlf: None,
            core_eol: None,
        };
        let result = crlf.apply(input, &source);
        assert!(matches!(result, FilterResult::Passthrough));
    }

    #[test]
    fn test_filter_list_load_text_file() {
        let tmp = test_tmp("filter_list_text");
        let workdir = &tmp;
        let git_dir = tmp.join(".git");

        // Create .gitattributes with text attribute
        fs::write(workdir.join(".gitattributes"), "*.txt text\n").unwrap();

        let list = FilterList::load(
            &git_dir,
            Some(workdir),
            "hello.txt",
            FilterMode::ToOdb,
            None,
        );
        assert!(list.contains("crlf"));
    }

    #[test]
    fn test_filter_list_load_binary_file() {
        let tmp = test_tmp("filter_list_binary");
        let workdir = &tmp;
        let git_dir = tmp.join(".git");

        fs::write(workdir.join(".gitattributes"), "*.bin binary\n").unwrap();

        let list = FilterList::load(
            &git_dir,
            Some(workdir),
            "image.bin",
            FilterMode::ToOdb,
            None,
        );
        assert!(list.is_empty());
    }

    #[test]
    fn test_filter_list_load_ident() {
        let tmp = test_tmp("filter_list_ident");
        let workdir = &tmp;
        let git_dir = tmp.join(".git");

        fs::write(workdir.join(".gitattributes"), "*.c ident\n").unwrap();

        let list = FilterList::load(
            &git_dir,
            Some(workdir),
            "main.c",
            FilterMode::ToWorktree,
            Some(OID::from_hex("aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d").unwrap()),
        );
        assert!(list.contains("ident"));
    }

    #[test]
    fn test_filter_list_apply_clean() {
        let tmp = test_tmp("filter_list_apply_clean");
        let workdir = &tmp;
        let git_dir = tmp.join(".git");

        fs::write(workdir.join(".gitattributes"), "*.txt text ident\n").unwrap();

        let list = FilterList::load(
            &git_dir,
            Some(workdir),
            "readme.txt",
            FilterMode::ToOdb,
            None,
        );

        let input = b"Version: $Id: abc123 $\r\nHello\r\n";
        let output = list.apply(input);

        // Both ident clean ($Id: ... $ → $Id$) and CRLF clean (CRLF → LF)
        assert_eq!(output, b"Version: $Id$\nHello\n");
    }

    #[test]
    fn test_filter_list_apply_smudge() {
        let tmp = test_tmp("filter_list_apply_smudge");
        let workdir = &tmp;
        let git_dir = tmp.join(".git");

        fs::write(
            workdir.join(".gitattributes"),
            "*.txt text eol=crlf ident\n",
        )
        .unwrap();

        let oid = OID::from_hex("aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d").unwrap();
        let list = FilterList::load(
            &git_dir,
            Some(workdir),
            "readme.txt",
            FilterMode::ToWorktree,
            Some(oid),
        );

        let input = b"Version: $Id$\nHello\n";
        let output = list.apply(input);

        // CRLF smudge (LF → CRLF) + ident smudge ($Id$ → $Id: hex $)
        let s = std::str::from_utf8(&output).unwrap();
        assert!(s.contains("$Id: aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d $"));
        assert!(s.contains("\r\n"));
    }

    #[test]
    fn test_eol_lf_attribute() {
        let tmp = test_tmp("filter_eol_lf");
        let workdir = &tmp;
        let git_dir = tmp.join(".git");

        fs::write(workdir.join(".gitattributes"), "*.txt text eol=lf\n").unwrap();

        // On smudge with eol=lf, no CRLF conversion
        let list = FilterList::load(
            &git_dir,
            Some(workdir),
            "readme.txt",
            FilterMode::ToWorktree,
            None,
        );
        // eol=lf means no LF→CRLF on smudge
        assert!(!list.contains("crlf"));
    }

    #[test]
    fn test_autocrlf_config() {
        let tmp = test_tmp("filter_autocrlf");
        let workdir = &tmp;
        let git_dir = tmp.join(".git");

        // Set core.autocrlf=true in config
        fs::write(
            git_dir.join("config"),
            "[core]\n\tautocrlf = true\n",
        )
        .unwrap();
        // No gitattributes needed — autocrlf applies globally

        let list = FilterList::load(
            &git_dir,
            Some(workdir),
            "readme.txt",
            FilterMode::ToOdb,
            None,
        );
        assert!(list.contains("crlf"));
    }

    #[test]
    fn test_empty_filter_list() {
        let tmp = test_tmp("filter_empty");
        let git_dir = tmp.join(".git");

        // No .gitattributes, no config
        let list = FilterList::load(&git_dir, Some(&tmp), "test.txt", FilterMode::ToOdb, None);
        assert!(list.is_empty());
        assert_eq!(list.len(), 0);

        // Apply should return input unchanged
        let input = b"hello world\n";
        assert_eq!(list.apply(input), input);
    }

    #[test]
    fn test_multiple_ident_markers() {
        let input = b"$Id$ and $Id$ again";
        let oid = OID::from_hex("aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d").unwrap();
        let result = ident_smudge(input, Some(&oid));
        match result {
            FilterResult::Applied(data) => {
                let s = std::str::from_utf8(&data).unwrap();
                // Both markers should be expanded
                assert_eq!(
                    s.matches("$Id: aaf4c61ddcc5e8a2dabede0f3b482cd9aea9434d $").count(),
                    2
                );
            }
            FilterResult::Passthrough => panic!("expected Applied"),
        }
    }

    #[test]
    fn test_ident_clean_multiple() {
        let input = b"$Id: abc $ and $Id: def $";
        let result = ident_clean(input);
        match result {
            FilterResult::Applied(data) => {
                assert_eq!(data, b"$Id$ and $Id$");
            }
            FilterResult::Passthrough => panic!("expected Applied"),
        }
    }
}
